File: app/src/main/java/com/clockmods/weather/WeatherRepository.java
package com.clockmods.weather;

import android.content.Context;
import android.content.SharedPreferences;

import com.clockmods.background.ClockPreferences;

import org.json.JSONObject;

import static com.clockmods.weather.WeatherModels.WeatherDetail;
import static com.clockmods.weather.WeatherModels.WeatherDisplayData;

public final class WeatherRepository {
    private static final String PREFS = "weather_cache";
    private final Context appContext;
    private final SharedPreferences preferences;
    public WeatherRepository(Context context) {
        appContext = context.getApplicationContext();
        preferences = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }
    public WeatherDisplayData getCached(String source, String locationId) {
        String value = preferences.getString("data", null);
        if (value == null) return null;
        try {
            JSONObject json = new JSONObject(value);
            String cachedSource = json.optString("source", "automatic");
            if (!source.equals(cachedSource)) return null;
            if ("manual".equals(source) && !locationId.equals(json.optString("locationId"))) return null;
            // Cached text is in the language it was fetched in; drop it after a language switch so
            // the UI never shows stale, wrong-language weather.
            if (!currentLanguage().equals(json.optString("lang"))) return null;
            return new WeatherDisplayData(json.optString("locationId"), json.optString("city"),
                    json.optString("district"), json.optString("text"), json.optString("icon"),
                    json.optString("temperature"), json.optLong("updatedAt"), readDetail(json));
        } catch (Exception ignored) { return null; }
    }
    public void save(WeatherDisplayData data, String source) {
        try {
            JSONObject json = new JSONObject();
            json.put("locationId", data.locationId).put("city", data.city).put("district", data.district)
                    .put("text", data.text).put("icon", data.icon).put("temperature", data.temperature)
                    .put("updatedAt", data.updatedAt).put("source", source).put("lang", currentLanguage());
            if (data.detail != null) json.put("detail", writeDetail(data.detail));
            preferences.edit().putString("data", json.toString()).apply();
        } catch (Exception ignored) { }
    }
    private String currentLanguage() {
        return new ClockPreferences(appContext).getClockLanguage();
    }
    private static WeatherDetail readDetail(JSONObject json) {
        JSONObject detail = json.optJSONObject("detail");
        if (detail == null) return null;
        return new WeatherDetail(
                nullable(detail, "feelsLike"), nullable(detail, "humidity"),
                nullable(detail, "windDir"), nullable(detail, "windScale"),
                nullable(detail, "precip"), nullable(detail, "warning"),
                nullable(detail, "aqiValue"), nullable(detail, "aqiCategory"));
    }
    private static JSONObject writeDetail(WeatherDetail detail) throws org.json.JSONException {
        JSONObject json = new JSONObject();
        json.put("feelsLike", detail.feelsLike).put("humidity", detail.humidity)
                .put("windDir", detail.windDir).put("windScale", detail.windScale)
                .put("precip", detail.precip).put("warning", detail.warning)
                .put("aqiValue", detail.aqiValue).put("aqiCategory", detail.aqiCategory);
        return json;
    }
    private static String nullable(JSONObject json, String key) {
        return json.isNull(key) ? null : json.optString(key, null);
    }
}

File: app/src/main/java/com/clockmods/weather/WeatherRepository.kt
package com.clockmods.weather

import android.content.Context
import android.content.SharedPreferences
import com.clockmods.background.ClockPreferences
import org.json.JSONObject

class WeatherRepository(context: Context) {
    private val appContext = context.applicationContext
    private val preferences: SharedPreferences = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun getCached(source: String, locationId: String): WeatherModels.WeatherDisplayData? {
        val value = preferences.getString("data", null) ?: return null
        return try {
            val json = JSONObject(value)
            if (source != json.optString("source", "automatic")) return null
            if (source == "manual" && locationId != json.optString("locationId")) return null
            if (currentLanguage() != json.optString("lang")) return null
            WeatherModels.WeatherDisplayData(json.optString("locationId"), json.optString("city"), json.optString("district"), json.optString("text"), json.optString("icon"), json.optString("temperature"), json.optLong("updatedAt"), readDetail(json))
        } catch (_: Exception) { null }
    }

    fun save(data: WeatherModels.WeatherDisplayData, source: String) {
        try {
            val json = JSONObject()
            putNullable(json, "locationId", data.locationId)
            putNullable(json, "city", data.city)
            putNullable(json, "district", data.district)
            putNullable(json, "text", data.text)
            putNullable(json, "icon", data.icon)
            putNullable(json, "temperature", data.temperature)
            json.put("updatedAt", data.updatedAt).put("source", source).put("lang", currentLanguage())
            data.detail?.let { json.put("detail", writeDetail(it)) }
            preferences.edit().putString("data", json.toString()).apply()
            appContext.sendBroadcast(android.content.Intent("com.clockmods.widget.WEATHER_CHANGED").setComponent(android.content.ComponentName(appContext.packageName, "com.clockmods.widget.update.WidgetDataChangedReceiver")))
        } catch (_: Exception) { }
    }

    private fun currentLanguage() = ClockPreferences(appContext).getClockLanguage()
    private fun readDetail(json: JSONObject): WeatherModels.WeatherDetail? = json.optJSONObject("detail")?.let { detail -> WeatherModels.WeatherDetail(nullable(detail, "feelsLike"), nullable(detail, "humidity"), nullable(detail, "windDir"), nullable(detail, "windScale"), nullable(detail, "precip"), nullable(detail, "warning"), nullable(detail, "aqiValue"), nullable(detail, "aqiCategory")) }
    private fun writeDetail(detail: WeatherModels.WeatherDetail) = JSONObject().also { json ->
        putNullable(json, "feelsLike", detail.feelsLike)
        putNullable(json, "humidity", detail.humidity)
        putNullable(json, "windDir", detail.windDir)
        putNullable(json, "windScale", detail.windScale)
        putNullable(json, "precip", detail.precip)
        putNullable(json, "warning", detail.warning)
        putNullable(json, "aqiValue", detail.aqiValue)
        putNullable(json, "aqiCategory", detail.aqiCategory)
    }
    private fun putNullable(json: JSONObject, key: String, value: String?) {
        if (value == null) json.remove(key) else json.put(key, value)
    }
    private fun nullable(json: JSONObject, key: String): String? =
        if (json.isNull(key)) null else json.optString(key)
    companion object { private const val PREFS = "weather_cache" }
}
